### ISSUE TITLE
The Normalizing Flow Network

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11848]
The Normalizing Flow Network<br>
<br>
https://ift.tt/WgCKfco<br>
<br>
<br>
<br>

<div><div><div><div><article><h1>The Normalizing Flow Network</h1>
<p>The Normalizing Flow Network (NFN) is a normalizing-flow based regression model, great at modelling complex conditional densities.
Look at our recent <a href="https://arxiv.org/abs/1907.08982">paper</a> on noise regularization for conditional density estimation for some results of using the NFN on real-world and benchmark regression datasets.</p>
<p>Here I’ll explain the structure of the NFN and go through some of the math.
Implementations can be found in our Open-Source <a href="https://github.com/freelunchtheorem/Conditional_Density_Estimation">Python package</a> as well as in my <a href="https://github.com/siboehm/NormalizingFlowNetwork">repo</a>.</p>
<h2>Normalizing Flows</h2>
<p>Let’s start with the density model, the final part of the NFN.
We are using Normalizing Flows as a way to parametrize the distribution we are outputting.
This expressive family can model characteristics of distributions like heteroscedasticity<label></label><span>This means that the variance of \(Y\) is not constant over \(X\). In this dataset for example, the variance of Y is higher for \(X\leq0\).<img alt="hetero datasets" src="https://siboehm.com/assets/img/nfn/hetero_noise.png"></span> and fat tails<label></label><span>The <a href="https://en.wikipedia.org/wiki/Cauchy_distribution">Cauchy distribution</a> is one example of a fat-tailed distribution. Compared to a Gaussian, it has outliers that are far more likely (So far so that it’s mean cannot even be defined).</span>.</p>
<p><img alt="Normalizing Flow" src="https://siboehm.com/assets/img/nfn/normalizing_flow.png"></p>
<p>A Normalizing Flow turns a simple base distribution \(p(x)\) (e.g. a Uniform or Gaussian distribution) into a more complex distribution \(\hat{p}(y)\).
This is done by applying an invertible, parametric function \(f(x)\) to samples \(x\) drawn from the base distribution, resulting in samples \(y\) from the transformed distribution.</p>
<p>Let’s start with a simple example: Take a random variable \(X\) distributed according to a Normal distribution with zero mean and uniform variance.
What is the distribution of, say \(Y = f(X) = \frac{1}{3} \cdot X\)?
Since we divide every sample \(x \sim p(X)\) by 3 to get a sample \(y \sim \hat{p}(Y)\), we can expect the distribution of Y to be more narrow, with most samples being closer to the mean than in the base distribution.</p>
<p>Now how to we calculate the density of Y?
A sample \(y \sim \hat{p}(Y)\) is generated by sampling \(x \sim p(X)\) and then transforming it to get \(f(x) = y\).
For a given sample y, we can get the x that generated it by using the inverse transformation \(f^{-1}(y) = 3y = x\). We can then calculate the likelihood of this x: \(p(f^{-1}(y)) = p(x) = p(3y)\).
Yet that is not equal to the likelihood of our sample y.
Since the density has been compressed, samples in the neighbourhood of y will be more likely than they had been under the (more stretched out) standard normal distribution.
This can be seen in the plot, the distribution \(p(3y)\) is degenerate and doesn’t integrate to 1.
Therefore we need to normalize.
This can be done by multiplying with the absolute derivative of the inverse of our function, \(|\frac{df^{-1}}{dy}| = |\frac{df}{dx}|^{-1}\).</p>
<p><img alt="transformed distribution" src="https://siboehm.com/assets/img/nfn/trans_dist.png"></p>
<p>This makes intuitive sense:</p>
<ul>
<li>\(\vert\frac{df}{dx}\vert \leq 1\): The function has a compressing effect. Samples closer together will be more likely so we need to scale up the transformed density.</li>
<li>\(\vert\frac{df}{dx}\vert &gt; 1\): The function expands the original density. Samples closer together will be less likely and we need to scale down to normalize.</li>
</ul>
<p>For multivariate distributions, the absolute derivate of the inverse function becomes the absolute determinant of the jacobian of the inverse function: \(\mid\det \frac{\partial f^{-1}}{\partial \mathbf{y}}\mid\).<label></label><span>While it’s a little harder to get an intuition for why that is the case, consider affine transformations in higher dimensions. They are easier to visualize than the general case since their Jacobian is just the transformation matrix. The determinant of a matrix is nothing but the amount of volume change resulting from the affine transformation defined by the matrix. For the 2-D case, check out this <a href="https://ncase.me/matrix/">applet</a>. Image the ‘M’ as a square. Then change the transformation matrix and see how the volume of the square is proportional to the determinant of the matrix.</span>
This leads us to the main formula of Normalizing Flows which tells us how to normalize a distribution transformed by \(f(\mathbf{x})\):</p>
\[\hat{p}(\mathbf{y}) = p(f^{-1}(\mathbf{y}))\Big|\det\frac{\partial f^{-1}}{\partial \mathbf{y}}\Big| = p(\mathbf{x})\Big|\det\frac{\partial f}{\partial \mathbf{x}}\Big|^{-1}\]
<p>This equation can be derived using the change-of-variable theorem. \(f(\mathbf{x})\) has to be continuously differentiable and invertible. Many of such functions have been proposed.
For the NFN we implemented three of them that we have found to be both expressive and fast to compute: affine, planar and radial flows.<label></label><span>We picked those three because unlike some of the recent flows like MAF or Real NVP they work especially well for low-dimensional densities.</span></p>
<h3>Affine Flows</h3>
<p>The example I used in the beginning of this post is an instance of the Affine Normalizing Flow:</p>
\[\mathbf{a},\mathbf{b} \in \mathbb{R}^d \quad f(\mathbf{x}) = \exp(\mathbf{a}) \odot \mathbf{x} + \mathbf{b}\]
<p>They are the simplest Normalizing Flow and can be used to shift and scale distributions.<label></label><span>They are not able to split a distribution into two modes though. To model multimodal distributions, we will combine them with other Normalizing Flows.</span>
Here we use an affine flow to transform a standard normal distribution:</p>
<p><img alt="affine Normalizing Flow" src="https://siboehm.com/assets/img/nfn/affine_flow.png"></p>
<p>The initial 2-D Gaussian is shifted and compressed, yet stays Gaussian. \(\mathbf{a}\) controls the scaling of the distribution and \(\mathbf{b}\) is the shift vector.</p>
<h3>Planar Flows</h3>
<p>Planar Flows compress and expand densities around a hyperplane:</p>
\[\mathbf{w}, \mathbf{u} \in \mathbb{R}^d, b \in \mathbb{R} \quad f(\mathbf{x}) = \mathbf{x} + \mathbf{u} \tanh(\mathbf{w}^\text{T}\mathbf{x} + b)\]
<p><img alt="planar flow" src="https://siboehm.com/assets/img/nfn/planar_flow.png"></p>
<p>The Gaussian base distribution is split into two halfs in this example.
\(b\) and \(\mathbf{w}\) define the hyperplane whereas \(\mathbf{u}\) specifies the direction and strength of the expansion.<label></label><span>Side note: While equation (3) is invertible using the right constraints, no closed form inverse is know. Therefore in practice the inverse can only be approximated.</span></p>
<h3>Radial Flows</h3>
<p>Radial flows compress and expand around a reference point:</p>
\[\gamma \in \mathbb{R}^d, \alpha, \beta \in \mathbb{R} \quad f(\mathbf{x}) = \mathbf{x} + \frac{\alpha\beta(\mathbf{x} - \gamma)}{\alpha + ||\mathbf{x}-\gamma||}\]
<p><img alt="radial flow" src="https://siboehm.com/assets/img/nfn/radial_flow.png"></p>
<p>\(\gamma\) is the reference point, \(\alpha, \beta\) specify how strongly to compress / expand.</p>
<p>Both planar and radial flows are only invertible functions if their parameters are correctly constrained.
How to constrain each flow and how to calculate the inverse log determinant of the Jacobian is explained in the documentation for my <a href="https://freelunchtheorem.github.io/Conditional_Density_Estimation/docs/html/density_estimator/nf.html">code</a> and in <a href="https://arxiv.org/abs/1505.05770">Rezende and Mohamed (2015)</a>.</p>
<p>Applying one Normalizing Flow to a distribution results in another distribution.
Therefore we can stack multiple flows to get even more expressive transformed distributions<label></label><span>Two consecutive planar flows transforming an initially Gaussian distribution: <img alt="chain of planar flows" src="https://siboehm.com/assets/img/nfn/planar_flow_chain.png"></span>.</p>
<h2>The Normalizing Flow Network</h2>
<p>Now let’s combine the Normalizing Flows with a neural network that allows us to model conditional distributions.
This concept is similar to a Mixture Density Network.<label></label><span>For the MDN the density model is a Mixture of Gaussians, for the NFN it’s a Normalizing Flow.</span>
The NFN is used to predict the conditional distribution \(p(y|x)\).<label></label><span>This is a generalization of regression / classification. In a regular regression task for example, one predicts y given x as \(\mathbb{E}(y|x)\). We generalize this by predicting a full distribution instead of just a point estimate.</span>
Say \(x\) is a job title and we’re trying to predict \(y\), the salary associated with it.<label></label><span>For example for p(wage | job=’lawyer’), this conditional distribution is strongly bimodal as you can see <a href="https://www.nalp.org/startingsalarydistributionclassof2009">here</a>. Some lawyers are much better off than others.</span>
The NFN now works as follows:</p>
<ol>
<li>A neural net takes x as it’s input and outputs the parameters of a stack of Normalizing Flows, e.g. two planar flows and one affine flow (The exact setup of the flow is a hyperparameter of the NFN).</li>
<li>The Normalizing Flows now define a transformation of a base Gaussian distribution into a more complex distribution, the conditional \(p(y\vert x)\).</li>
<li>The model is then trained to minimize the negative log likelihood of the data:</li>
</ol>
\[L(\theta) = -\sum_{n=1}^N \log p(\mathbf{y_n}\mid NN_{\theta}(\mathbf{x_n}))\]
<p><img alt="Normalizing Flow network" src="https://siboehm.com/assets/img/nfn/normalizing_flow_network.png"></p>
<p>As a small detail, the Normalizing Flows are actually inverted in the NFN and point from the transformed distribution to the base distribution.
This is mainly done to speed up the process, as the flows we use here we developed for posterior density estimation in variational inference by <a href="https://arxiv.org/abs/1505.05770">Rezende and Mohamed (2015)</a>. In their original form they are fast at sampling and slow at likelihood computation. For CDE we need the opposite.</p>
<p>The NFN fitted on a conditional version of the two moons dataset where \(X\) determines the rotation of \(Y\) around the origin:
<img alt="conditional two moons" src="https://siboehm.com/assets/img/nfn/conditional_two_moons.png"></p>
<h2>Conclusion</h2>
<p>Normalizing flows serve as a powerful, novel way of parametrizing distributions.
In the tests in our <a href="https://arxiv.org/abs/1907.08982">paper</a> we compared the NFN against the Mixture Density Network and the Kernel Mixture Network, among others.
The NFN shows favourable performance on datasets that are strongly non-Gaussian, e.g. datasets that exhibit skewness and heavy tails.</p>
<p>You can try the NFN and other conditional density estimators on your own data using our <a href="https://github.com/freelunchtheorem/Conditional_Density_Estimation">Python package</a> (TensorFlow 1.X) or my <a href="https://github.com/siboehm/NormalizingFlowNetwork">repo</a> (TensorFlow 2.0).</p>
<h2>Further Info</h2>
</article></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via siboehm <br>
<br>
October 21, 2024 at 05:39PM
